modules/alloydb/main.tf

```diff
@@ -0,0 +1,372 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+locals {
+  prefix = var.prefix == null ? "" : "${var.prefix}-"
+  # has_replicas = try(length(var.replicas) > 0, false)
+  is_regional = var.availability_type == "REGIONAL" ? true : false
+  # secondary instance type is aligned with cluster type unless apply is targeting a promotion, in that
+  # case cluster will be 'primary' while instance still 'secondary'.
+  secondary_cluster_name  = coalesce(var.secondary_cluster_name, "${var.cluster_name}-sec")
+  secondary_instance_name = coalesce(var.secondary_name, "${var.name}-sec")
+  secondary_instance_type = try(var.cross_region_replication.promote_secondary && google_alloydb_cluster.secondary[0].cluster_type == "SECONDARY" ? "SECONDARY" : google_alloydb_cluster.secondary[0].cluster_type, null)
+  users = {
+    for k, v in coalesce(var.users, {}) :
+    k => {
+      name     = k
+      password = try(v.type, "ALLOYDB_BUILT_IN") == "ALLOYDB_BUILT_IN" ? try(random_password.passwords[k].result, v.password) : null
+      roles    = v.roles
+      type     = coalesce(v.type, "ALLOYDB_BUILT_IN")
+    }
+  }
+}
+
+resource "google_alloydb_cluster" "primary" {
+  project          = var.project_id
+  cluster_id       = "${local.prefix}${var.cluster_name}"
+  database_version = var.database_version
+  deletion_policy  = var.deletion_policy
+  labels           = var.labels
+  location         = var.location
```

Comment:
You're missing a few fields (e.g. `display_name`, `annotations`)

Reply:
Added both display_name (both clusters) and annotations (same applies for all the instances and cluster). Thanks for the spot, I decided not to use `network` since it is deprecated in favor of `network_config` as well as arguments for setting up a cluster from a backup (restore_*). wdyt?

Reply:
Lgtm